### PR TITLE
feat(edge): secret validation

### PR DESCRIPTION
**Pull request by caspiano**
Adds `PlaceOS::Model::Edge#check?(secret)` for the purpose of validating secrets of connecting edge nodes.

### CODE REVIEW

**Review by dukeraphaelng — APPROVED** [review]
looks good

File: src/placeos-models/edge.cr [hunk 1]
```diff
@@ -29,10 +33,12 @@ module PlaceOS::Model
       foreign_key: "edge_id",
     )
 
-    before_save :encrypt!
+    def valid?(test : String) : Bool
+      Encryption.decrypt(string: secret, id: salt, level: ENCRYPTION_LEVEL) == test
+    end
 
     def encrypt
```
Comment (KimBurgess) [inline src/placeos-models/edge.cr:40]: It looks like this encrypts that secret only. To make this a little clearer (particularly if there's a need to encrypt multiple attributes down the track, what about `encrypt_secret` to make this clearer? Ditto for the decrypt pair.

File: src/placeos-models/edge.cr [hunk 2]
```diff
@@ -29,10 +33,12 @@ module PlaceOS::Model
       foreign_key: "edge_id",
     )
 
-    before_save :encrypt!
+    def valid?(test : String) : Bool
+      Encryption.decrypt(string: secret, id: salt, level: ENCRYPTION_LEVEL) == test
```
Comment (KimBurgess) [inline src/placeos-models/edge.cr:37]: May be safer to encrypt the input string and compare, rather than popping a decrypted value into mem.

Comment (caspiano) [reply src/placeos-models/edge.cr:37]: i thought so too, but the iv is random. i'll look into how we've implemented encryption

File: src/placeos-models/edge.cr (hunk 1 above)
Comment (caspiano) [reply src/placeos-models/edge.cr:40]: `valid_secret?` might be clearer?

Comment (caspiano) [reply src/placeos-models/edge.cr:40]: oh you mean specifically regarding the attribute. yep all for that.

Comment (KimBurgess) [reply src/placeos-models/edge.cr:40]: Yep, sorry - the comment attached to the incorrect line.

File: src/placeos-models/utilities/encryption.cr [hunk 3]
```diff
@@ -27,30 +27,36 @@ module PlaceOS::Encryption
...
+  def self.check?(encrypted : String, test : String, id : String, level : Level) : Bool
+    return true if level == Level::None && encrypted == test
+    return false unless is_encrypted?(encrypted)
+
+    salt, iv, cipher_text = extract_context(encrypted)
+
+    _, key = generate_key(id: id, level: level, salt: salt)
+
+    encrypted_test = encrypt_data(data: test, key: key, salt: salt, iv: ::Base64.decode(iv))
+    encrypted_test[:cipher_text] == cipher_text
```
Comment (KimBurgess) [inline src/placeos-models/utilities/encryption.cr:42]:
```suggestion
    Crypto::Subtle.constant_time_compare encrypted_test[:cipher_text], cipher_text
```
See https://crystal-lang.org/api/0.35.1/Crypto/Subtle.html#constant_time_compare(x,y):Bool-class-method.

**Review by KimBurgess — COMMENTED** [review]
General meta-question for this - is there every read these secrets?

If not, an alternative approach could be to just use `Crypto::Bcrypt::Password` from the std lib. Specifically [`#verify`](https://crystal-lang.org/api/0.35.1/Crypto/Bcrypt/Password.html#verify(password:String):Bool-instance-method).

File: src/placeos-models/utilities/encryption.cr (hunk 3 above)
Comment (caspiano) [reply src/placeos-models/utilities/encryption.cr]: For now we'll just allow admins to view the secrets.

In the future, we could just shoot the secret back with the create and warn the user they'll need to note the secret for deployment of their edge nodes.

**Pull request merged**